### PR TITLE
Migrate `staticlib-dylib-linkage` `run-make` test to rmake

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

I'm quite sure this has the same issue as the one brought up in [this discussion](https://github.com/rust-lang/rust/pull/128407#discussion_r1702439172), so I elected to keep the ignore MSVC.

try-job: aarch64-apple
try-job: x86_64-gnu-llvm-17
try-job: armhf-gnu

### CODE REVIEW

File: tests/run-make/staticlib-dylib-linkage/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        .input("foo.rs")
+        .crate_type("staticlib")
+        .print("native-static-libs")
+        .arg("-Zstaticlib-allow-rdylib-deps")
+        .run()
+        .assert_stderr_contains("note: native-static-libs: ")
+        .stderr_utf8();
+    let re = regex::Regex::new(r#"note: native-static-libs:\s*(.+)"#).unwrap();
+    let libs = re.find(&libs).unwrap().as_str().trim();
+    let libs: &Vec<&str> = &libs[libs.find("-L").unwrap()..libs.len()].split(' ').collect();
```
Comment (jieyouxu) [inline tests/run-make/staticlib-dylib-linkage/rmake.rs:25]: Nit: this seemed a bit hard to understand, maybe something like

```suggestion
let (_, library_search_paths) = libs.split_once("-L").unwrap();
let library_search_paths = library_search_paths.split(' ').collect::<Vec<_>>();
```

File: tests/run-make/staticlib-dylib-linkage/rmake.rs
Comment (jieyouxu) [inline tests/run-make/staticlib-dylib-linkage/rmake.rs:1]: Suggestion: we should add a quick comment about what this test attempts to exercise. Namely, a basic smoke test to check that rustc supports linking to rust dylib with `--crate-type staticlib` as taught by #106560.

**Review by jieyouxu — APPROVED** [review]
Thanks, this seems reasonable, just a few nits.

**Comment by Oneirical** [comment]
All review remarks applied. @rustbot review

**Comment by jieyouxu** [comment]
Thanks, r=me after CI is green
@bors delegate+ rollup=iffy

**Pull request merged**